core/src/main/java/org/springframework/security/oauth2/server/authorization/web/OAuth2TokenEndpointFilter.java

```diff
@@ -29,16 +44,56 @@
 
 /**
  * @author Joe Grandja
+ * @author Madhu Bhat
  */
 public class OAuth2TokenEndpointFilter extends OncePerRequestFilter {
-	private Converter<HttpServletRequest, Authentication> authorizationGrantConverter;
+	private final String DEFAULT_TOKEN_URI = "/oauth2/token";
+	private Converter<HttpServletRequest, OAuth2AuthorizationCodeAuthenticationToken> authorizationGrantConverter;
 	private AuthenticationManager authenticationManager;
 	private OAuth2AuthorizationService authorizationService;
+	private RequestMatcher uriMatcher;
+
+	/**
+	 * Constructs an {@code OAuth2TokenEndpointFilter} using the provided parameter.
+	 *
+	 * @param authorizationService the authorization service implementation
+	 */
+	public OAuth2TokenEndpointFilter(OAuth2AuthorizationService authorizationService) {
+		Assert.notNull(authorizationService, "authorizationService cannot be null");
+		this.authorizationGrantConverter = new OAuth2AuthorizationCodeAuthenticationTokenConverter();
+		this.authenticationManager = new OAuth2ClientAuthenticationManager();
+		this.uriMatcher = new AntPathRequestMatcher(DEFAULT_TOKEN_URI, HttpMethod.POST.name());
```

Comment:
In addition to the path, the `RequestMatcher` should also match on parameter `grant_type=authorization_code`

Reply:
So in case `grant_type!=authorization_code` shouldn't the auth server send back `unsupported_grant_type` as an error response as per [5.2 Error Response](https://tools.ietf.org/html/rfc6749#section-5.2)? Having the grant_type param check in the requestMatcher would just ignore the request in case the request has an invalid `grant_type`. Is that fine?

Reply:
> So in case grant_type!=authorization_code shouldn't the auth server send back unsupported_grant_type as an error response

Yes, you are correct. So let's just match on the  `grant_type ` parameter and return `unsupported_grant_type` if there isn't a match on the value of the parameter.

Reply:
Just ensuring that we're on the same page, we don't need the grant_type added to the requestMatcher, but validated as part of the request validation, correct?

Reply:
Sorry for not being clear. The `RequestMatcher` should also match on the existence of the `grant_type` parameter name (not value). The parameter value will be used after to determine if it can be converted

Reply:
@jgrandja just one doubt though. We can match with RequestMatcher if `grant_type` was passed as a query param using a regex with `RegexRequestMatcher`. However, if the `grant_type` was passed within the posted form data, the same cannot be matched with `RequestMatcher` right? Your thoughts on that please?

Reply:
Take a look at `OAuth2AuthorizationCodeGrantFilter.matchesAuthorizationResponse()`, which demonstrates more advanced request matching. **NOTE:** Although `OAuth2AuthorizationCodeGrantFilter` does not use a `RequestMatcher`, the `matchesAuthorizationResponse()` method can be referenced as a `RequestMatcher`.

Reply:
@jgrandja as per [section-4.1.3](https://tools.ietf.org/html/rfc6749#section-4.1.3), `grant_type` is a required parameter for an access token request and as per [section-5.2](https://tools.ietf.org/html/rfc6749#section-5.2) which is referenced by [section-4.1.4](https://tools.ietf.org/html/rfc6749#section-4.1.4) of access token response, the response should have error as
>invalid_request
               The request is missing a required parameter

Doesn't this mean that the `RequestMatcher` shouldn't include existence of `grant_type`, but rather be validated post the request being matched and validated along with the rest of the required params for the existence of `grant_type` param?

Reply:
@kratostaine Thank you for reading the details in the spec! And for keeping me in check :) You are correct. If the `grant_type`  parameter is not provided then an `OAuth2Error` should be returned with `invalid_request`. So no need to use `grant_type` in the `RequestMatcher`.